CoreFoundation/PlugIn.subproj/CFBundle_Main.c

```diff
@@ -79,12 +79,20 @@ static CFBundleRef _CFBundleGetMainBundleAlreadyLocked(void) {
         CFStringRef str = NULL;
         CFURLRef executableURL = NULL, bundleURL = NULL;
         _initedMainBundle = true;
+#ifdef __ANDROID__
```

Comment:
This seems to be setting the bundle URL to `$CFFIXED_USER_HOME` or (a) `$TMPDIR`, which is highly surprising for downstream users — this is supposed to be a directory path where the application (as 'application' is defined by the platform) is storing read-only resource files.

Some quick research suggests this would be covered by the `AAssetManager` API on Android; can we go the distance for that? Why isn't the default (creating a bundle out of the path the executable is in) correct on this platform?

Reply:
I would stick with the CF behavior of using the current path in this case.

Reply:
There is no useful equivalent to a “bundle directory” on Android as you would think of it on iOS. App resources are held compressed and the only means of access to them is through a stream which isn’t useful for example in specifying a “cacerts.pem” file for lib curl to use for accessing https: sites. The nearest equivalent is the “cacheDir” or “internalDataPath” as it is made available in C which you could consider the "Documents" and “tmp” dir from iOS all rolled into one. This is the only directory an application knows the absolute path for and has write access to as far as I know.

For bundle resources all you can do is copy your resource out of it’s stream into this directory on the Java side of things, as in, for example, [this code](https://github.com/SwiftJava/swift-android-kotlin/blob/master/app/src/main/java/com/example/user/myapplication/MainActivity.kt#L48) and press this directory into service as the bundle directory so the following code works in Swift:

```Swift
    func setCacheDir( cacheDir: String? ) {
        setenv( "TMPDIR", cacheDir!, 1 )

        // Required for SSL to work
        let pemFile = Bundle.main.path(forResource:"cacert", ofType:"pem")
        setenv( "URLSessionCertificateAuthorityInfoFile", pemFile, 1 )
```
Using the AAssetManager api would increase the surface area of foundation exposed to Android unnecessarily and would require support from the Java end of things to get hold of the activity instance to get hold of the AssetManager etc which isn’t straightforward. Far easier to just pass the path of the cacheDir to the Swift code as that is all it’s interested in and keep things simple. The only dependency foundation has at the moment is the __android_log_print function.

Reply:
`AAssetManager` is also only available from a prohibitively high API version compared to what we'd like to support. I couldn't find the exact version in a quick google search but the AAssetManager was only added to the NDK in 2016 from what I could see.

It's a valid point though, what happens if you just enter no path whatsoever? Logically this should default to a path relative to the currently running app's... bundle directory... but I have a vague suspicion that Android doesn't do that and instead just throws an exception / assertion failure of sorts. Have you tested this recently @johnno1962 ?

Reply:
The default "bundle directory" in this PR is /data/local/tmp. How do you mean “specify no path"? I agree we really don’t want to go down the C route. The AAssetManager api is just a limited version of what is available from Java.

Reply:
I appreciate the discussion. My confusion is: how does the code under it, which defines the bundle path to be essentially dirname(executable), not work for the Android case?

Reply:
It comes up with /system/bin which is no use in practical terms.